test/unit/signal.spec.js

```diff
@@ -1,36 +1,167 @@
+      })
+      it('throws an error if listener is not a function', () => {
+        ['foo', undefined, 42].forEach((value) => {
+          expect(() => { memoizedSignal.attach(value) }).to.throw()
+        })
+      })
+      it('throws if there is no initial value provided', () => {
+        expect(() => new MemoizedSignal()).to.throw()
+      })
+      it('detaches listener when returned function is invoked', () => {
+        const detach = memoizedSignal.attach(spies.one)
+
+        sinon.assert.calledOnce(spies.one)
+        spies.one.reset()
+        detach()
+        memoizedSignal.dispatch()
+        sinon.assert.notCalled(spies.one)
+      })
+    })
+
+    describe('dispatch()', () => {
+      it('returns nothing and does not fail', () => {
+        expect(memoizedSignal.dispatch()).to.equal(undefined)
+      })
+
+      describe('with previously attached listeners', () => {
+        function expectCallCount (obj) {
+          for (let name in obj) {
+            expect(spies[name]).to.have.callCount(obj[name])
+          }
+        }
+
+        it('fires attached listeners in the same order they were attached', () => {
+          memoizedSignal.attach(spies.one)
+          memoizedSignal.attach(spies.two)
+          memoizedSignal.attach(spies.three)
+          memoizedSignal.dispatch()
+
+          expectCallCount({one: 2, two: 2, three: 2})
+
+          sinon.assert.callOrder(spies.one, spies.two, spies.three)
+        })
+
+        it('does not fire detached listeners', () => {
+          memoizedSignal.attach(spies.one)()
+          memoizedSignal.attach(spies.two)
+          memoizedSignal.dispatch()
+
+          expectCallCount({one: 1, two: 2})
+        })
+
+        it('fires reattached listeners', () => {
+          memoizedSignal.attach(spies.one)()
+          memoizedSignal.dispatch()
+
+          expectCallCount({one: 1})
+
+          memoizedSignal.attach(spies.one)
+          memoizedSignal.dispatch()
+
+          expectCallCount({one: 3})
+        })
+
+        it('fires same listener attached twice two times', () => {
+          const detachSpyOneA = memoizedSignal.attach(spies.one)
+          memoizedSignal.attach(spies.one)
+          spies.one.reset() // reset call count increment cause by memoization behaviour
+          memoizedSignal.dispatch()
+          expectCallCount({one: 2})
+
+          detachSpyOneA()
+          memoizedSignal.dispatch()
+
+          expectCallCount({one: 3})
+        })
+
+        it('passes given arguments to the listeners', () => {
+          const args = ['foo', 42, {}]
+
+          memoizedSignal.attach(spies.one)
+          sinon.assert.calledWithExactly(spies.one, initialValue)
+
+          memoizedSignal.dispatch(...args)
+          sinon.assert.calledWithExactly(spies.one, ...args)
+
+          memoizedSignal.dispatch()
+          sinon.assert.calledWithExactly(spies.one)
+        })
+
+        it('memoizes last dispatched value', () => {
```

Comment:
Be mor descriptive: 'calls listener with most recently dispatched value'.

Please also add a test case for firing with the initial value provided to the constructor.